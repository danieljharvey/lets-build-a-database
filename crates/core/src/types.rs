use std::{fmt::Display, hash::Hash};

#[derive(Debug, PartialOrd, PartialEq, Eq, Ord, Hash, Clone)]
pub struct Column {
    pub name: String,
    pub table_alias: Option<TableAlias>,
}

impl Display for Column {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        match &self.table_alias {
            Some(table_alias) => {
                write!(f, "{}.{}", table_alias, self.name)
            }
            None => {
                write!(f, "{}", self.name)
            }
        }
    }
}

impl std::convert::From<&str> for Column {
    fn from(name: &str) -> Column {
        Column {
            name: name.to_string(),
            table_alias: None,
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum Expr {
    ColumnComparison {
        column: Column,
        op: Op,
        literal: serde_json::Value,
    },
}

#[derive(Debug, PartialEq)]
pub enum Op {
    Equals,
    GreaterThan,
    GreaterThanOrEqual,
    LessThan,
    LessThanOrEqual,
}

#[derive(Debug, PartialEq)]
pub struct Join {
    #[allow(clippy::struct_field_names)]
    pub join_type: JoinType,
    pub left_from: Box<Query>,
    pub right_from: Box<Query>,
    pub left_column_on: Column,
    pub right_column_on: Column,
}

#[derive(Debug, PartialEq)]
pub struct TableName(pub String);

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone)]
pub struct TableAlias(pub String);

impl Display for TableAlias {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
        write!(f, "{}", self.0)
    }
}

#[derive(Debug, PartialEq)]
pub struct From {
    pub table_name: TableName,
    pub table_alias: Option<TableAlias>,
}

#[derive(Debug, PartialEq)]
pub struct Filter {
    pub from: Box<Query>,
    pub filter: Expr,
}

#[derive(Debug, PartialEq)]
pub struct Project {
    pub from: Box<Query>,
    pub fields: Vec<Column>,
}

#[derive(Debug, PartialEq)]
pub enum JoinType {
    Inner,
    LeftOuter,
}

#[derive(Debug, PartialEq)]
pub struct Limit {
    pub from: Box<Query>,
    pub limit: u64,
}

#[derive(Debug, PartialEq)]
pub enum Query {
    From(From),
    Filter(Filter),
    Join(Join),
    Project(Project),
    Limit(Limit),
}

#[derive(Debug, PartialEq, Clone)]
pub struct Row {
    pub items: Vec<serde_json::Value>,
}

impl Row {
    pub fn get_column(&self, column: &Column, schema: &Schema) -> Option<&serde_json::Value> {
        let index = schema.get_index_for_column(column)?;

        self.items.get(index)
    }

    pub fn extend(&mut self, row: Row) {
        self.items.extend(row.items);
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct Schema {
    pub columns: Vec<Column>,
}

impl Schema {
    pub fn get_index_for_column(&self, column: &Column) -> Option<usize> {
        self.columns
            .iter()
            .enumerate()
            .find(|(_, column_name)| column_name == &column)
            .map(|(i, _)| i)
    }

    pub fn extend(&mut self, schema: Schema) {
        self.columns.extend(schema.columns);
    }
}

pub struct QueryStep {
    pub schema: Schema,
    pub rows: Vec<Row>,
}

impl QueryStep {
    // reconstruct JSON output
    pub fn to_json(&self) -> serde_json::Value {
        let mut output_rows = vec![];

        for row in &self.rows {
            let mut output_row = serde_json::Map::new();
            for column in &self.schema.columns {
                let value = row.get_column(column, &self.schema).unwrap();
                output_row.insert(column.to_string(), value.clone());
            }
            output_rows.push(serde_json::Value::Object(output_row));
        }
        serde_json::Value::Array(output_rows)
    }
}
