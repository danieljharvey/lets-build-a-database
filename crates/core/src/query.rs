mod filter;
mod from;
mod join;
mod order_by;
mod project;

use crate::types::{Limit, OrderBy};

use super::types::QueryStep;
use super::types::{Column, Filter, From, Join, Project, Query};

#[derive(Debug)]
pub enum QueryError {
    ColumnNotFoundInSchema { column_name: Column },
    IndexNotFoundInSchema { index: usize },
    FilterError(filter::FilterError),
}

pub fn run_query(query: &Query) -> Result<QueryStep, QueryError> {
    match query {
        Query::From(From {
            table_name,
            table_alias,
        }) => Ok(from::table_scan(table_name, table_alias.as_ref())),
        Query::Filter(Filter { from, filter }) => {
            let QueryStep {
                schema,
                rows,
                mut cost,
            } = run_query(from)?;

            let mut filtered_rows = vec![];

            for row in rows {
                cost.increment_rows_processed();
                if filter::apply_predicate(&row, &schema, filter)? {
                    filtered_rows.push(row);
                }
            }

            Ok(QueryStep {
                schema,
                rows: filtered_rows,
                cost,
            })
        }
        Query::Project(Project { from, fields }) => {
            let QueryStep {
                schema,
                rows,
                mut cost,
            } = run_query(from)?;

            let mut projected_rows = vec![];

            for row in &rows {
                cost.increment_rows_processed();
                projected_rows.push(project::project_fields(row, &schema, fields)?);
            }

            let schema = project::project_schema(&schema, fields)?;

            Ok(QueryStep {
                schema,
                rows: projected_rows,
                cost,
            })
        }
        Query::Limit(Limit { limit, from }) => {
            let QueryStep {
                schema,
                mut rows,
                cost,
            } = run_query(from)?;
            let size: usize = (*limit).try_into().unwrap();

            rows.truncate(size);

            Ok(QueryStep { schema, rows, cost })
        }
        Query::Join(Join {
            left_from,
            right_from,
            join_type,
            on,
        }) => {
            let QueryStep {
                schema: left_schema,
                rows: left_rows,
                cost: mut left_cost,
            } = run_query(left_from)?;

            let QueryStep {
                schema: right_schema,
                rows: right_rows,
                cost: right_cost,
            } = run_query(right_from)?;

            left_cost.extend(&right_cost);

            join::hash_join(
                left_rows,
                &left_schema,
                right_rows,
                &right_schema,
                on,
                join_type,
                left_cost,
            )
        }
        Query::OrderBy(OrderBy {
            from,
            order_by_exprs,
        }) => {
            let QueryStep {
                schema,
                rows,
                mut cost,
            } = run_query(from)?;

            let rows = order_by::order_by(rows, &schema, order_by_exprs, &mut cost);

            Ok(QueryStep { schema, rows, cost })
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::{parser::parse, run_query};

    #[test]
    fn test_query_select_animals() {
        let query = parse("SELECT * FROM animal").unwrap();

        let result = run_query(&query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_query_select_horse() {
        let query = parse("select * from animal where animal_name = 'horse'").unwrap();
        let result = run_query(&query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_query_projection() {
        let query = parse("select animal_name from animal where animal_name = 'horse'").unwrap();
        let result = run_query(&query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_select_horse_and_species() {
        let query = parse(
            r"
        select * from animal 
        join species 
            on species_id 
        where animal_name = 'horse'",
        )
        .unwrap();
        let result = run_query(&query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_select_species_and_animals() {
        let query = parse(
            r"
        select * from species 
          join animal on species_id
        where
          species_id >= 3 
    ",
        )
        .unwrap();
        let result = run_query(&query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_select_species_and_animals_left_outer() {
        let query = parse(
            r"
        select * from species 
          left outer join animal on species_id
        where
          species_id = 3 
    ",
        )
        .unwrap();
        let result = run_query(&query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_select_album() {
        let query = parse(
            r"
        select * from Album 
        where Title = 'Jagged Little Pill'
    ",
        )
        .unwrap();
        let result = run_query(&query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_select_album_and_artist() {
        let query = parse(
            r"
        select * from Album 
          join Artist on ArtistId
        where
          ArtistId = 82
    ",
        )
        .unwrap();
        let result = run_query(&query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_select_track_album_and_artist() {
        let query = parse(
            r"
        select Name, Title, artist.Name from Track
          join Album on AlbumId
          join Artist as artist on ArtistId
        where
          ArtistId = 82
        limit 10
    ",
        )
        .unwrap();
        let result = run_query(&query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_select_filter_with_column_reference() {
        let query = parse(
            r"
        select * from Album 
        where AlbumId = (ArtistId + 1 + 1) - 1 
    ",
        )
        .unwrap();
        let result = run_query(&query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_select_order_by_name_limit_5() {
        let query = parse(
            r"
        select * from Album
        order by Title
        limit 5
    ",
        )
        .unwrap();
        let result = run_query(&query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }

    #[test]
    fn test_select_order_by_artist_id_and_title_with_limit() {
        let query = parse(
            r"
        select ArtistId, Title from Album
        order by ArtistId, Title
        limit 4
    ",
        )
        .unwrap();
        let result = run_query(&query).unwrap();

        insta::assert_json_snapshot!(result.to_json());
        insta::assert_debug_snapshot!(result.cost);
    }
}
