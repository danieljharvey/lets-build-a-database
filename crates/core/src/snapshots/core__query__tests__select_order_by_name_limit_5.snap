---
source: crates/core/src/query.rs
expression: result.to_json()
---
[
  {
    "AlbumId": 208,
    "ArtistId": 136,
    "Title": "(1997) Black Light Syndrome"
  },
  {
    "AlbumId": 156,
    "ArtistId": 50,
    "Title": "...And Justice For All"
  },
  {
    "AlbumId": 257,
    "ArtistId": 179,
    "Title": "20th Century Masters - The Millennium Collection: The Best of Scorpions"
  },
  {
    "AlbumId": 296,
    "ArtistId": 230,
    "Title": "A Copland Celebration, Vol. I"
  },
  {
    "AlbumId": 94,
    "ArtistId": 90,
    "Title": "A Matter of Life and Death"
  }
]
